src/libraries/System.Private.CoreLib/src/System/Number.Dragon4.cs

```diff
@@ -100,203 +101,123 @@ public static unsafe void Dragon4Single(float value, int cutoffNumber, bool isSi
-            // 0.69 is chosen because 0.69 + log10(2) is less than one by a reasonable epsilon that will account for any floating point error.
-            //
-            // We want to set digitExponent to floor(log10(v)) + 1
-            //      v = mantissa * 2^exponent
-            //      log2(v) = log2(mantissa) + exponent;
-            //      log10(v) = log2(v) * log10(2)
-            //      floor(log2(v)) = mantissaHighBitIdx + exponent;
-            //      log10(v) - log10(2) < (mantissaHighBitIdx + exponent) * log10(2) <= log10(v)
-            //      log10(v) < (mantissaHighBitIdx + exponent) * log10(2) + log10(2) <= log10(v) + log10(2)
-            //      floor(log10(v)) < ceil((mantissaHighBitIdx + exponent) * log10(2)) <= floor(log10(v)) + 1
-            const double Log10V2 = 0.30102999566398119521373889472449;
-            int digitExponent = (int)(Math.Ceiling(((int)(mantissaHighBitIdx) + exponent) * Log10V2 - 0.69));
-
-            // Divide value by 10^digitExponent.
-            if (digitExponent > 0)
+            if (len > 1)
             {
-                // The exponent is positive creating a division so we multiply up the scale.
-                scale.MultiplyPow10((uint)(digitExponent));
+                mid = descaledValue.GetBlock(1);
             }
-            else if (digitExponent < 0)
+            if (len > 2)
             {
-                // The exponent is negative creating a multiplication so we multiply up the scaledValue, scaledMarginLow and scaledMarginHigh.
+                high = descaledValue.GetBlock(2);
+            }
+            if (len > 3)
+            {
+                uint idk = descaledValue.GetBlock(3);
+                throw new Exception("TODO handle this somehow" + idk);
+            }
 
-                BigInteger.Pow10((uint)(-digitExponent), out BigInteger pow10);
+/*          UInt128 valueUInt128 = 0;
+            valueUInt128 += high;
+            valueUInt128 = valueUInt128 << 32;
+            valueUInt128 += mid;
+            valueUInt128 = valueUInt128 << 32;
+            valueUInt128 += low;
+            Console.WriteLine(valueUInt128.ToString());
+*/ 
 
-                scaledValue.Multiply(ref pow10);
-                scaledMarginLow.Multiply(ref pow10);
+            // Since the scale in the Decimal type is representing dividing by 10^e, we want to invert decimalScale
+            uint decimalScale = 28 - ((uint)state.digitExponent - 1);
+            Debug.Assert(decimalScale <= 28 && decimalScale > 0);
+            return (low, mid, high, decimalScale);
 
-                if (pScaledMarginHigh != &scaledMarginLow)
-                {
-                    BigInteger.Multiply(ref scaledMarginLow, 2, out *pScaledMarginHigh);
-                }
-            }
+        }
 
-            bool isEven = (mantissa % 2) == 0;
-            bool estimateTooLow = false;
+        private unsafe ref struct Dragon4State
```

Comment:
This is the state that needs to persist from the first half to the second half of the Dragon4 algorithm. We are hijacking the output of the first half to use for our conversion.